packages/fx-core/tests/component/generator/generator.test.ts

```diff
@@ -45,39 +55,110 @@ describe("Generator utils", () => {
+  setTools(tools);
+  const ctx = createContextV3();
+  const sandbox = sinon.createSandbox();
+  const tmpDir = path.join(__dirname, "tmp");
+
+  afterEach(async () => {
+    sandbox.restore();
+  });
+
+  it("fetch sample url with tag error", async () => {
+    sandbox.stub(fetchSampleUrlWithTagAction, "run").throws(new Error("test"));
+    try {
+      await Generator.generateSample("bot-sso", tmpDir, ctx);
+    } catch (e) {
+      assert.equal(e.name, "FetchSampleUrlWithTagError");
+    }
+  });
+
+  it("fetch sample zip from url error", async () => {
+    sandbox.stub(fetchSampleUrlWithTagAction, "run").resolves();
+    sandbox.stub(fetchZipFromUrlAction, "run").throws(new Error("test"));
+    try {
+      await Generator.generateSample("bot-sso", tmpDir, ctx);
+    } catch (e) {
+      assert.equal(e.name, "FetchZipFromUrlError");
+    }
+  });
+
+  it("template fallback error", async () => {
+    sandbox.stub(fetchTemplateUrlWithTagAction, "run").throws(new Error("test"));
+    sandbox.stub(fetchTemplateZipFromLocalAction, "run").throws(new Error("test"));
+    try {
+      await Generator.generateTemplate("bot", "ts", tmpDir, ctx);
+    } catch (e) {
+      assert.equal(e.name, "TemplateZipFallbackError");
+    }
+  });
+
+  it("unzip error", async () => {
+    sandbox.stub(fetchTemplateUrlWithTagAction, "run").resolves();
+    sandbox.stub(fetchZipFromUrlAction, "run").resolves();
+    sandbox.stub(fetchTemplateZipFromLocalAction, "run").resolves();
+    sandbox.stub(unzipAction, "run").throws(new Error("test"));
+    try {
+      await Generator.generateTemplate("bot", "ts", tmpDir, ctx);
+    } catch (e) {
+      assert.equal(e.name, "UnzipError");
+    }
   });
 });
 
-describe("Generator happy path", () => {
+describe("Generator happy path", async () => {
   const tools = new MockTools();
   setTools(tools);
   const context = createContextV3();
+  const sandbox = sinon.createSandbox();
   const tmpDir = path.join(__dirname, "tmp");
 
   afterEach(async () => {
+    sandbox.restore();
     if (await fs.pathExists(tmpDir)) {
       await fs.rm(tmpDir, { recursive: true });
     }
   });
+
   it("external sample", async () => {
     const sampleName = "bot-proactive-messaging-teamsfx";
     await Generator.generateSample(sampleName, tmpDir, context);
-    const files = await fs.readdir(path.join(tmpDir, sampleName));
+    const files = await fs.readdir(tmpDir);
+    assert.isTrue(files.length > 0);
+    assert.isTrue(files.includes(".fx"));
+  });
+
+  it("template", async () => {
+    const templateName = "bot";
+    const language = "ts";
+    sandbox
+      .stub(fetchTemplateUrlWithTagAction, "run")
+      .callsFake(async (context: GeneratorContext) => {
+        context.zipUrl =
+          "https://github.com/hund030/TemplatePackerDemo/releases/download/templates%400.1.0/bot_notification_ts_function_http.zip";
+      });
+    await Generator.generateTemplate(templateName, language, tmpDir, context);
```

Comment:
Seems this test downloads the template and unzip it? It's better to mock the api call.